pkg/container-collection/options.go

```diff
@@ -41,68 +38,94 @@ import (
 	return func(cc *ContainerCollection) error {
-		dockercli, err := dockerclient.NewClientWithOpts(dockerclient.FromEnv)
+		// Already running containers
+		runtimeClient, err := containerutils.NewContainerRuntimeClient("")
 		if err != nil {
 			return err
 		}
 
-		// Already running containers
-		containers, err := dockercli.ContainerList(context.Background(),
-			dockertypes.ContainerListOptions{
-				All: true,
-			})
+		containers, err := runtimeClient.GetContainers("")
 		if err != nil {
 			return err
 		}
+
 		for _, container := range containers {
-			res, err := dockercli.ContainerInspect(context.Background(), container.ID)
-			if err != nil {
-				log.Errorf("failed to inspect container %s: %s", container.ID, err)
+			if !container.Running {
+				log.Debugf("Runtime enricher: ignoring non running container: %s",
+					container.ID)
 				continue
 			}
-			if !res.State.Running {
+
+			pid, err := runtimeClient.PidFromContainerID(container.ID)
+			if err != nil {
+				log.Errorf("Runtime enricher: failed to get PID of container %s: %s",
+					container.ID, err)
 				continue
 			}
-			if res.State.Pid == 0 {
-				log.Errorf("failed to inspect container %s: container pid is 0", container.ID)
+			if pid == 0 {
+				log.Errorf("Runtime enricher: ignoring container with zero PID: %s",
+					container.ID)
 				continue
 			}
-			pid := res.State.Pid
+
+			name := strings.TrimPrefix(container.Name, "/")
+
 			cc.initialContainers = append(cc.initialContainers,
 				&pb.ContainerDefinition{
-					Id:  container.ID,
-					Pid: uint32(pid),
+					Id:   container.ID,
+					Pid:  uint32(pid),
+					Name: name,
+
+					// Some gadgets require the namespace and pod name to be set
+					Namespace: "default",
+					Podname:   name,
 				})
 		}
 
 		// Future containers
 		cc.containerEnrichers = append(cc.containerEnrichers, func(container *pb.ContainerDefinition) bool {
-			filter := dockerfilters.NewArgs()
-			filter.Add("id", container.Id)
-			containers, err := dockercli.ContainerList(context.Background(),
-				dockertypes.ContainerListOptions{
-					All:     true,
-					Filters: filter,
-				})
+			// Container already enriched
+			if container.Name != "" && container.Namespace != "" && container.Podname != "" {
+				return true
+			}
+
+			containers, err := runtimeClient.GetContainers(container.Id)
 			if err != nil {
-				log.Errorf("failed to list container %s: %s", container.Id, err)
+				log.Debugf("Runtime enricher: failed to get container %s: %s",
+					container.Id, err)
 				return true
 			}
-			if len(containers) == 1 {
-				if len(containers[0].Names) > 0 {
-					container.Name = strings.TrimPrefix(containers[0].Names[0], "/")
+
+			switch len(containers) {
+			case 0:
+				// Drop containers not visible by the container runtime. For
+				// instance, pause containers are not exposed by containerd.
+				log.Debugf("Runtime enricher: no containers found with ID %s",
+					container.Id)
+				return false
```

Comment:
We are not including the pause containers in our list of containers when using CRI-O and Contained CRIs because they are not part of the response of `CRI.ListContainers()`. However, when using the Docker Engine API, we are indeed including them because they are part of the response of `DockerClient.ContainerList()`. Should we ignore them too in the Docker case? For instance, by discarding containers with the command "/pause"?

Reply:
I understand your point if the new approach is to support several runtimes. However, if it is the case of a pause container, it will end up being added to the container list without the metadata. It means that the `list-containers` command will have a strange empty line 😐. Yes, there are several ways we can avoid that, like enriching the view of the `list-containers` command to also show the ID or using the ID as the default name in case it is not enriched by any runtime, etc. But my question is, is it really useful for users to have the pause containers in the list?

Reply:
I decided to keep them. Let's try to develop the local-gadget no Kubernetes-centric and treat all containers identically.